Simple worker heartbeat to detect dead workers

tomtor opened the pull request:
See #818 

Add a simple worker heartbeat (every 60s) for reliable scavenging (remove as active after 3 minutes).

Also show a worker immediately when it receives a task and do not wait for the first batch of games to be completed.

Finally replace `utils/scavenge.py` with server logic.

ppigazzini commented on worker/worker.py:
```diff
@@ -13,6 +13,7 @@
 import time
 import traceback
 import uuid
+import threading
```
According `isort` this is the order:
```python
import sys
import threading
import time
```

tomtor replied:
Fixed

ppigazzini commented:
@tomtor DEV and workers updated.

tomtor commented:
@ppigazzini Thanks, but for some reason the task_id passed from the worker is None.

Looking into it.

tomtor commented:
@ppigazzini I was distracted by a local hard disk crash, but local testing seems to work now....

vondele commented on worker/worker.py:
```diff
@@ -336,6 +365,10 @@ def main():
         "unique_key": str(uuid.uuid4()),
     }
 
+    # Start hart beat
```
nit: hart beat -> heartbeat

tomtor replied:
fixed

ppigazzini commented:
DEV and workers updated.

> @ppigazzini I was distracted by a local hard disk crash, but local testing seems to work now....

!!!

tomtor commented:
I cleaned up (UPPERCASE for globals) and changed that a worker is visible from the start with `0/0/0` stats so we do not have to wait for the first game results to show up.

tomtor commented:
@ppigazzini I modified the server scavenging. I expect that in general `utils/scavenge.py` is no longer needed, so we could remove it from the crontab and see how that goes....

ppigazzini commented:
@tomtor DEV and workers updated, `utils/scavenge.py` commented out in `crontab`.

tomtor commented:
@ppigazzini https://dfts-0.pigazzini.it/tests/view/5f8f45655916ba61557f0642 gives an error

Can you please paste the error log?

ppigazzini commented:
Log error
```
Oct 21 11:22:24 vps1572612706 pserve[28639]:   File "/home/usr00/fishtest/fishtest/fishtest/rundb.py", line 737, in sync_update_task
Oct 21 11:22:24 vps1572612706 pserve[28639]:     fishtest.stats.stat_util.update_SPRT(self.get_results(run, False), sprt)
Oct 21 11:22:24 vps1572612706 pserve[28639]:   File "/home/usr00/fishtest/fishtest/fishtest/rundb.py", line 465, in get_results
Oct 21 11:22:24 vps1572612706 pserve[28639]:     results["crashes"] += stats["crashes"]
Oct 21 11:22:24 vps1572612706 pserve[28639]: KeyError: 'crashes'
```

I reloaded a clean state for DEV and workers to double check the status.

tomtor commented:
@ppigazzini Thanks, should be fixed now!

ppigazzini commented:
@tomtor clean state reloaded for DEV and workers :)

ppigazzini commented:
@tomtor I'm starting several workers with a bash script and killing them with `pkill -fe "python3 worker.py"`

Log error:
```
Oct 21 12:13:21 vps1572612706 pserve[30997]: 2020-10-21 12:13:21,201 ERROR [waitress][waitress-3] Exception while serving /api/beat
Oct 21 12:13:21 vps1572612706 pserve[30997]: Traceback (most recent call last):
...
Oct 21 12:13:21 vps1572612706 pserve[30997]:     response = getattr(inst, attr)()
Oct 21 12:13:21 vps1572612706 pserve[30997]:   File "/home/usr00/fishtest/fishtest/fishtest/api.py", line 240, in beat
Oct 21 12:13:21 vps1572612706 pserve[30997]:     if self.task_id() is not None:
Oct 21 12:13:21 vps1572612706 pserve[30997]:   File "/home/usr00/fishtest/fishtest/fishtest/api.py", line 93, in task_id
Oct 21 12:13:21 vps1572612706 pserve[30997]:     return int(self.request.json_body["task_id"])
Oct 21 12:13:21 vps1572612706 pserve[30997]: TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

tomtor commented:
@ppigazzini Thanks, should be fixed now

ppigazzini commented:
@tomtor reloaded.

ppigazzini commented:
@tomtor logs clean, dead workers with signal are removed immediately, dead workers closing the terminal are removed after the 5 minutes timeout.

Tonight I will test the case of network drop worker side for over 5 minutes.

tomtor commented:
> after the 5 minutes timeout.

It will be after minimal 3 and maximal 4 minutes...

ppigazzini commented on worker/worker.py:
```diff
@@ -109,6 +110,32 @@ def get_rate():
+                headers={"Content-type": "application/json"},
+                timeout=HTTP_TIMEOUT,
+            )
+            req = json.loads(req.text)
```
Perhaps better to use something like this to avoid a confused worker output during a network drop:
```python
            try:
                req = json.loads(req.text)
                print(req)
            except Exception as e:
                sys.stderr.write("Exception from calling heartbeat:\n")
                print(e)
```

ppigazzini commented:
@tomtor LGTM also during network drops, IMO we should clean up the output with a try/except.

tomtor commented:
@ppigazzini Thanks, Added your code.
Not sure if we should print the full exception or just a short message that it failed.

tomtor commented:
@ppigazzini Regarding migration. I suggest after merging installing the server code with a timeout of 30 minutes first (patch by hand). The new server code should be compatible with old workers.

After the workers have been upgraded we can lower that again to 3 minutes.

ppigazzini commented:
@tomtor DEV and workers updated.

I simply copied this code, no preference about the information showed during an exception, a mild preference about uniformity in the codebase:
https://github.com/glinscott/fishtest/blob/bba962e5c118ae54178f456f7c12e45a55f19b75/worker/games.py#L697-L700

I have been running since yesterday also a worker on master, no problems in my tests (w/ USTC, though).

To avoid misunderstanding the parameter to set temporary to 30 by hand after the commit is this:
```python
def scavenge_tasks(scavenge=True, minutes=5):
```

tomtor commented on fishtest/fishtest/rundb.py:
```diff
@@ -315,10 +324,15 @@ def flush_buffers(self):
-        old = datetime.utcnow() - timedelta(minutes=110)
+        # print("scavenge ", run["_id"])
+        dead_task = False
+        old = datetime.utcnow() - timedelta(minutes=3)
```
@ppigazzini Adjust timeout here: `minutes=30`

I expect that we remove `utils/scavenge.py` from the crontab, otherwise we need to change it there too

ppigazzini replied:
@tomtor thanks for the info, I will comment out `scavenge.py` on PROD crontab

tomtor commented:
> a mild preference about uniformity in the codebase:

Yes, see https://github.com/glinscott/fishtest/pull/821/commits/1fda5dd53551bc27d45134b0905703257c2a37a8 for uniformity

tomtor commented:
@ppigazzini For me it's OK to squash and merge and install on prod. Let me know when I can squash.

ppigazzini commented:
@squash :)

squash commented:
> @squash :)

Yes?

ppigazzini commented:
@noobpwnftw please update the fleet at your convenience :)

ppigazzini commented:
PROD updated, timeout raised to 30 minutes by hand (server restarted), `scavenge.py` commented out in crontab.
Thank you @tomtor :)

ppigazzini commented:
> > @squash :)
> 
> Yes?

@squash Typo, sorry :)

ppigazzini commented:
@tomtor `fish.exit` doesn't work anymore :(

```
Score of New-560c776397 vs Base-560c776397: 64 - 79 - 57  [0.463] 200
  Task updated successfully in 0.190487s
Server told us task is no longer needed
Wait 1 seconds before upload of PGN...
Uploading compressed PGN of 184152 bytes
Task exited
Send heartbeat... Pleased to hear from you...
Send heartbeat... Pleased to hear from you...
Send heartbeat... Pleased to hear from you...
Send heartbeat... Pleased to hear from you...
```

ppigazzini commented:
#831 should fix the problem
